benches/fraud.rs

```diff
@@ -0,0 +1,399 @@
+                    },
+                )
+                .join::<(), _, _, _>(
+                    &enriched_transactions,
+                    |(_cc_num, _ts), (pw_avg, pm_avg, freq), (tran, dem)| QueryResult {
+                        avg_spend_pw: *pw_avg,
+                        avg_spend_pm: *pm_avg,
+                        trans_freq_24: *freq as u32,
+                        category: tran.category,
+                        amt: tran.amt,
+                        state: dem.state,
+                        job: dem.job,
+                        unix_time: tran.unix_time,
+                        city_pop: dem.city_pop,
+                        merchant: tran.merchant,
+                        is_fraud: tran.is_fraud,
+                    },
+                );
+
+            (hdemographics, htransactions)
+        })
+        .unwrap();
+
+        Self {
+            dbsp,
+            hdemographics,
+            htransactions,
+        }
+    }
+
+    fn ingest_demographics(&mut self, path: &str) {
+        let mut dem_reader = CsvReader::from_path(path).unwrap();
+
+        println!("Ingesting demographics");
+        let dem_iter = dem_reader.deserialize::<Demographics>();
+        for record in dem_iter {
+            let record = record.unwrap();
+            // println!("Person: {record:?}");
+            self.hdemographics.push(record, 1);
+        }
+        self.dbsp.step().unwrap();
+    }
+
+    fn process_transactions<R: Read>(mut self, mut reader: CsvReader<R>, batch_size: usize) {
+        let start = Instant::now();
+        println!("Ingesting transactions");
+
+        let trans_iter = reader.deserialize::<Transaction>();
+        let chunks = trans_iter.chunks(batch_size);
+
+        let (tx, rx) = bounded(1);
+
+        let thread_handle = spawn(move || loop {
+            match rx.recv().unwrap() {
+                true => {
+                    let chunk_start = Instant::now();
+                    self.dbsp.step().unwrap();
+                    println!("compute: {}", chunk_start.elapsed());
+                }
+                false => {
+                    self.dbsp.kill().unwrap();
+                    break;
+                }
+            }
+        });
+
+        let mut total_size: usize = 0;
+        for chunk in chunks.into_iter() {
+            let chunk_start = Instant::now();
+            let mut batch: Vec<_> = chunk
+                .map(|record| {
+                    let transaction = record.unwrap();
+                    // println!("Transaction: {transaction:?}");
+                    (transaction, 1)
+                })
+                .collect();
+            self.htransactions.append(&mut batch);
+            println!("{total_size} parsing: {} ", chunk_start.elapsed());
+            total_size += batch_size;
+
+            tx.send(true).unwrap();
+        }
+
+        tx.send(false).unwrap();
+        thread_handle.join().unwrap();
+
+        println!("total time: {}", start.elapsed());
```

Comment:
how do you know the results are correct?

Reply:
I don't.  We need a test harness.